File: overlay/tests/test_getset.py
"""
test_getset.py - Test get/set operations, pipelining, and binary safety.
"""

import os
import sys
import unittest

sys.path.insert(0, os.path.dirname(__file__))
from memcached_harness import MemcachedServer


class TestGetSet(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.server = MemcachedServer().start()

    @classmethod
    def tearDownClass(cls):
        cls.server.stop()

    def setUp(self):
        self.mc = self.server.client()

    def tearDown(self):
        self.mc.close()

    def test_basic_set_get(self):
        self.assertEqual(self.mc.set("k_basic", "val_basic"), "STORED")
        res = self.mc.get("k_basic")
        self.assertEqual(res["k_basic"][0], "val_basic")

    def test_empty_value(self):
        self.assertEqual(self.mc.set("empty_k", ""), "STORED")
        res = self.mc.get("empty_k")
        self.assertEqual(res["empty_k"][0], "")

    def test_whitespace_in_value(self):
        val = "line1	line2   end"
        self.assertEqual(self.mc.set("complex_v", val), "STORED")
        res = self.mc.get("complex_v")
        self.assertEqual(res["complex_v"][0], val)

    def test_binary_characters_in_value(self):
        val = bytes([0, 1, 2, 255, 128, 64, 0, 10, 13, 0])
        self.assertEqual(self.mc.set("bin_val", val), "STORED")
        self.mc.send_cmd("get bin_val")
        line = self.mc.readline()
        parts = line.strip().split()
        self.assertEqual(parts[0], "VALUE")
        nbytes = int(parts[3])
        data = self.mc.read_bytes(nbytes)
        self.mc.read_bytes(2)  # consume

        self.mc.readline()  # consume END
        self.assertEqual(data, val)

    def test_pipelining(self):
        cmds = (
            b"set pipe1 0 0 4\r\nval1\r\nset pipe2 0 0 4\r\nval2\r\nget pipe1 pipe2\r\n"
        )
        self.mc.sock.sendall(cmds)
        self.assertEqual(self.mc.readline().strip(), "STORED")
        self.assertEqual(self.mc.readline().strip(), "STORED")
        res = {}
        while True:
            line = self.mc.readline().strip()
            if not line or line == "END":
                break
            parts = line.split()
            if len(parts) >= 4 and parts[0] == "VALUE":
                k = parts[1]
                nbytes = int(parts[3])
                data = self.mc.read_bytes(nbytes)
                self.mc.read_bytes(2)
                res[k] = data.decode("utf-8")
        self.assertEqual(res["pipe1"], "val1")
        self.assertEqual(res["pipe2"], "val2")


if __name__ == "__main__":
    unittest.main()

File: overlay/tests/test_getset.c
/*
 * test_getset.c - Test get/set operations, pipelining, and binary safety.
 */

#include "c_harness.h"

static int test_basic_set_get(void) {
  struct memcached_srv srv;
  struct memcached_client c;
  char resp[64];
  char val[64];

  TEST_ASSERT(srv_start(&srv, 0, NULL) == 0);
  TEST_ASSERT(client_connect(&c, srv.port) == 0);

  TEST_ASSERT(
      client_set(&c, "k_basic", "val_basic", 9, 0, 0, resp, sizeof(resp)) == 0);
  TEST_ASSERT_STR_EQ(resp, "STORED");

  TEST_ASSERT(client_get(&c, "k_basic", val, sizeof(val), NULL) == 0);
  TEST_ASSERT_STR_EQ(val, "val_basic");

  client_close(&c);
  srv_stop(&srv);
  return 0;
}

static int test_empty_value(void) {
  struct memcached_srv srv;
  struct memcached_client c;
  char resp[64];
  char val[64];

  TEST_ASSERT(srv_start(&srv, 0, NULL) == 0);
  TEST_ASSERT(client_connect(&c, srv.port) == 0);

  TEST_ASSERT(client_set(&c, "empty_k", "", 0, 0, 0, resp, sizeof(resp)) == 0);
  TEST_ASSERT_STR_EQ(resp, "STORED");

  TEST_ASSERT(client_get(&c, "empty_k", val, sizeof(val), NULL) == 0);
  TEST_ASSERT_STR_EQ(val, "");

  client_close(&c);
  srv_stop(&srv);
  return 0;
}

static int test_whitespace_in_value(void) {
  struct memcached_srv srv;
  struct memcached_client c;
  const char *payload = "line1\tline2   end";
  char resp[64];
  char val[128];

  TEST_ASSERT(srv_start(&srv, 0, NULL) == 0);
  TEST_ASSERT(client_connect(&c, srv.port) == 0);

  TEST_ASSERT(client_set(&c, "ws_key", payload, strlen(payload), 0, 0, resp,
                         sizeof(resp)) == 0);
  TEST_ASSERT_STR_EQ(resp, "STORED");

  TEST_ASSERT(client_get(&c, "ws_key", val, sizeof(val), NULL) == 0);
  TEST_ASSERT_STR_EQ(val, payload);

  client_close(&c);
  srv_stop(&srv);
  return 0;
}

static int test_binary_characters_in_value(void) {
  struct memcached_srv srv;
  struct memcached_client c;
  char binary_payload[10];
  char val[64];
  char resp[64];

  binary_payload[0] = 0;
  binary_payload[1] = 1;
  binary_payload[2] = 2;
  binary_payload[3] = (char)255;
  binary_payload[4] = (char)128;
  binary_payload[5] = 64;
  binary_payload[6] = 0;
  binary_payload[7] = 10;
  binary_payload[8] = 13;
  binary_payload[9] = 0;

  TEST_ASSERT(srv_start(&srv, 0, NULL) == 0);
  TEST_ASSERT(client_connect(&c, srv.port) == 0);

  TEST_ASSERT(client_set(&c, "bin_val", binary_payload, 10, 0, 0, resp,
                         sizeof(resp)) == 0);
  TEST_ASSERT_STR_EQ(resp, "STORED");

  TEST_ASSERT(client_get(&c, "bin_val", val, sizeof(val), NULL) == 0);
  TEST_ASSERT(memcmp(val, binary_payload, 10) == 0);

  client_close(&c);
  srv_stop(&srv);
  return 0;
}

static int test_pipelining(void) {
  struct memcached_srv srv;
  struct memcached_client c;
  const char *pipeline_req = "set pipe1 0 0 4\r\nval1\r\n"
                             "set pipe2 0 0 4\r\nval2\r\n"
                             "get pipe1 pipe2\r\n";
  char line[256];
  char v1[32], v2[32];
  int got_v1 = 0, got_v2 = 0;

  TEST_ASSERT(srv_start(&srv, 0, NULL) == 0);
  TEST_ASSERT(client_connect(&c, srv.port) == 0);

  TEST_ASSERT(client_send_str(&c, pipeline_req) == 0);

  TEST_ASSERT(client_readline(&c, line, sizeof(line)) > 0);
  TEST_ASSERT(strncmp(line, "STORED", 6) == 0);

  TEST_ASSERT(client_readline(&c, line, sizeof(line)) > 0);
  TEST_ASSERT(strncmp(line, "STORED", 6) == 0);

  while (client_readline(&c, line, sizeof(line)) > 0) {
    if (strncmp(line, "END", 3) == 0) {
      break;
    }
    if (strncmp(line, "VALUE pipe1 ", 12) == 0) {
      char crlf[2];
      TEST_ASSERT(client_read_bytes(&c, v1, 4) == 0);
      v1[4] = '\0';
      TEST_ASSERT(client_read_bytes(&c, crlf, 2) == 0);
      got_v1 = 1;
    } else if (strncmp(line, "VALUE pipe2 ", 12) == 0) {
      char crlf[2];
      TEST_ASSERT(client_read_bytes(&c, v2, 4) == 0);
      v2[4] = '\0';
      TEST_ASSERT(client_read_bytes(&c, crlf, 2) == 0);
      got_v2 = 1;
    }
  }

  TEST_ASSERT(got_v1 && got_v2);
  TEST_ASSERT_STR_EQ(v1, "val1");
  TEST_ASSERT_STR_EQ(v2, "val2");

  client_close(&c);
  srv_stop(&srv);
  return 0;
}

int test_getset_run(void) {
  RUN_TEST(test_basic_set_get);
  RUN_TEST(test_empty_value);
  RUN_TEST(test_whitespace_in_value);
  RUN_TEST(test_binary_characters_in_value);
  RUN_TEST(test_pipelining);
  return 0;
}
